transcoders.go

```diff
@@ -167,6 +168,93 @@ func onionBtS(b []byte) (string, error) {
 	return addr + ":" + strconv.Itoa(int(port)), nil
 }
 
+var TranscoderOnion3 = NewTranscoderFromFunctions(onion3StB, onion3BtS, nil)
+
+func onion3StB(s string) ([]byte, error) {
+	addr := strings.Split(s, ":")
+	if len(addr) != 2 {
+		return nil, fmt.Errorf("failed to parse onion addr: %s does not contain a port number.", s)
+	}
+
+	// onion address without the ".onion" substring
+	if len(addr[0]) != 56 {
+		return nil, fmt.Errorf("failed to parse onion addr: %s not a Tor onionv3 address. len == %d", s, len(addr[0]))
+	}
+	onionHostBytes, err := base32.StdEncoding.DecodeString(strings.ToUpper(addr[0]))
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode base32 onion addr: %s %s", s, err)
+	}
+
+	// onion port number
+	i, err := strconv.Atoi(addr[1])
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse onion addr: %s", err)
+	}
+	if i >= 65536 {
+		return nil, fmt.Errorf("failed to parse onion addr: %s", "port greater than 65536")
+	}
+	if i < 1 {
+		return nil, fmt.Errorf("failed to parse onion addr: %s", "port less than 1")
+	}
+
+	onionPortBytes := make([]byte, 2)
+	binary.BigEndian.PutUint16(onionPortBytes, uint16(i))
+	bytes := []byte{}
+	bytes = append(bytes, onionHostBytes[0:35]...)
+	bytes = append(bytes, onionPortBytes...)
+	return bytes, nil
+}
+
+func onion3BtS(b []byte) (string, error) {
+	addr := strings.ToLower(base32.StdEncoding.EncodeToString(b[0:35]))
+	port := binary.BigEndian.Uint16(b[35:37])
+	str := addr + ":" + strconv.Itoa(int(port))
+	return str, nil
+}
+
+var TranscoderGarlic64 = NewTranscoderFromFunctions(garlic64StB, garlic64BtS, garlicValidate)
+
+// i2p uses an alternate character set for base64 addresses. This returns an appropriate encoder.
+func garlicBase64Encoding() *base64.Encoding {
+	return base64.NewEncoding("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-~")
+}
+
+func garlic64StB(s string) ([]byte, error) {
+	// i2p base64 address
+	if len(s) < 516 || len(s) > 616 {
+		return nil, fmt.Errorf("failed to parse garlic addr: %s not an i2p base64 address. len: %d\n", s, len(s))
+	}
+	garlicHostBytes, err := garlicBase64Encoding().DecodeString(s)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode base64 i2p addr: %s %s", s, err)
+	}
+
+	bytes := []byte{}
+	bytes = append(bytes, garlicHostBytes...)
+
+	return bytes, nil
+}
+
+func garlic64BtS(b []byte) (string, error) {
+	addr := garlicBase64Encoding().EncodeToString(b)
+	return addr, nil
+}
+
+func garlicValidate(b []byte) error {
+	if len(b) < 386 {
+		return fmt.Errorf("failed to validate garlic addr: %s not an i2p base64 address. len: %d\n", b, len(b))
+	}
+	s, err := garlic64BtS(b)
+	if err != nil {
+		return fmt.Errorf("failed to decode base64 i2p addr: %s %s", s, err)
+	}
+	_, err = garlicBase64Encoding().DecodeString(s)
```

Comment:
Do we need to encode/decode? Validate's supposed to be a fast subset of `garlic64BtS` (one that only *checks* if the bytes are correct without actually turning them into a string). I *believe* simply checking the length would be sufficient.